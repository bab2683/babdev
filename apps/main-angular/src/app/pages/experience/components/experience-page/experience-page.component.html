<ng-container
  *ngIf="{
    isMobile: isMobile$ | async,
    pageData: pageData$ | async
  } as data"
>
  <babdev-page-with-bg
    [bgData]="data.pageData && data.pageData.bgData"
    [title]="'experience.TITLE' | translate | async"
  >
    <div class="experience" *ngIf="CV$ | async as CV">
      <babdev-card
        [withHover]="!data.isMobile"
        class="card"
        *ngFor="let exp of CV.sections.experiences.items"
      >
        <h2>
          {{ exp.title }}
        </h2>
        <h3 class="card__company">
          {{ exp.company }}
        </h3>
        <h4 class="card__period">
          {{ exp.dates.start | date: 'LLL yyyy' }} -
          {{
            exp.current
              ? ('common.TIME.PRESENT' | translate | async)
              : (exp.dates.end | date: 'LLL yyyy')
          }}
          ({{
            exp.duration.years
              ? exp.duration.years +
                ' ' +
                ((exp.duration.years > 1
                  ? 'common.TIME.YEAR.MULTIPLE'
                  : 'common.TIME.YEAR.SINGLE'
                )
                  | translate
                  | async) +
                ' '
              : ''
          }}{{
            exp.duration.months
              ? exp.duration.months +
                ' ' +
                ((exp.duration.months > 1
                  ? 'common.TIME.MONTH.MULTIPLE'
                  : 'common.TIME.MONTH.SINGLE'
                )
                  | translate
                  | async)
              : ''
          }})
        </h4>
        <div class="card__zones">
          <div class="card__main">
            <p>{{ exp.description | translate | async }}</p>
            <ul>
              <li *ngFor="let task of exp.tasks">
                {{ task | translate | async }}
              </li>
            </ul>
          </div>
          <div class="card__extra" *ngIf="exp.technologies.length">
            <p>
              <strong>{{ 'common.TECHNOLOGIES' | translate | async }}:</strong>
            </p>
            <ul>
              <li *ngFor="let tech of exp.technologies">
                <babdev-link-or-text
                  [key]="tech"
                  [source]="technologies"
                ></babdev-link-or-text>
              </li>
            </ul>
          </div>
        </div>
      </babdev-card>
    </div>
  </babdev-page-with-bg>
</ng-container>
