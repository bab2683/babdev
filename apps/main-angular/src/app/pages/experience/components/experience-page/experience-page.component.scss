@import "~@styleguide_funcional";

.experience {
  margin-top: px-to-rem(20);
}

.card {
  & + & {
    margin-top: px-to-rem(10);
  }

  &__company {
    margin-top: px-to-rem(5);
  }

  &__period {
    margin-top: px-to-rem(5);
  }

  &__zones {
    margin-top: px-to-rem(10);
  }

  &__main {
    padding-right: px-to-rem(16);
  }

  &__extra {
    margin-top: px-to-rem(10);
  }
}

:host-context(.isDesktop) {
  .card {
    &__zones {
      display: flex;
    }

    &__main {
      width: 70%;
    }

    &__extra {
      margin-top: 0;
      width: 30%;
    }
  }
}
