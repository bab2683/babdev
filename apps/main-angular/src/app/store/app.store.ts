import { routerReducer, RouterReducerState } from '@ngrx/router-store';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';

import { environment } from '@env/environment';
import {
  GlobalEffects,
  globalReducer,
  GlobalState,
  GLOBAL_FEATURE,
  ROUTER_FEATURE
} from './features';

export interface AppState {
  [GLOBAL_FEATURE]: GlobalState;
  [ROUTER_FEATURE]: RouterReducerState;
}

export const AppEffects: any[] = [GlobalEffects];

export const reducers: ActionReducerMap<AppState> = {
  [GLOBAL_FEATURE]: globalReducer,
  [ROUTER_FEATURE]: routerReducer
};

export const metaReducers: Array<MetaReducer<
  AppState
>> = !environment.production ? [] : [];
