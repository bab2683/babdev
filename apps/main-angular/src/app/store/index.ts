export * from './app.store';
export * from './features';
export * from './mocks';
