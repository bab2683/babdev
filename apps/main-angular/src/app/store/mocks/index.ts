export * from './app.store.mock';
