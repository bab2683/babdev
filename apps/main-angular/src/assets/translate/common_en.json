{
  "PAGES": {
    "HOME": "Home",
    "CONTACTS": "Contacts"
  },
  "TIME": {
    "YEAR": {
      "SINGLE": "year",
      "MULTIPLE": "years"
    },
    "MONTH": {
      "SINGLE": "month",
      "MULTIPLE": "months"
    },
    "DAY": {
      "SINGLE": "day",
      "MULTIPLE": "days"
    }
  }
}
