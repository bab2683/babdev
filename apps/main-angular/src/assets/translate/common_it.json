{
  "PAGES": {
    "HOME": "Home",
    "CONTACTS": "Contatti"
  },
  "TIME": {
    "YEAR": {
      "SINGLE": "anno",
      "MULTIPLE": "anni"
    },
    "MONTH": {
      "SINGLE": "mese",
      "MULTIPLE": "mesi"
    },
    "DAY": {
      "SINGLE": "giorno",
      "MULTIPLE": "giorni"
    }
  }
}
