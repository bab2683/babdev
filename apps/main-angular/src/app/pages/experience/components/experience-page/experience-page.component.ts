import { Component, OnInit } from '@angular/core';
import { Technologies } from '@babdev/layout';
import { RequestService } from '@babdev/request';
import { DictionaryLoader, TranslateService } from '@babdev/translate';
import { select, Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { map, take } from 'rxjs/operators';

import { CV } from '@models';
import { AppState, getIsMobileState } from '@store';
import { parseCVDates } from '@utils';

@Component({
  selector: 'babdev-experience-page',
  templateUrl: './experience-page.component.html',
  styleUrls: ['./experience-page.component.scss']
})
export class ExperiencePageComponent implements OnInit {
  public CV$: Observable<CV>;
  public isMobile$: Observable<boolean>;
  public technologies: Technologies;

  public bgPath: string = 'bg/experience';
  private dictionary: DictionaryLoader = { location: '/pages/experience/', name: 'experience' };

  constructor(
    private translateService: TranslateService,
    private req: RequestService,
    private store: Store<AppState>
  ) {}

  ngOnInit() {
    this.translateService.loadDictionary(this.dictionary);

    this.req
      .get<Technologies>({ cache: true, url: 'assets/data/technologies.json' })
      .pipe(take(1))
      .subscribe(({ data }) => (this.technologies = data));

    this.CV$ = this.req
      .get<CV>({
        cache: true,
        url: 'assets/data/cv.json'
      })
      .pipe(
        take(1),
        map(({ data }) => data),
        map<CV, CV>(parseCVDates)
      );

    this.isMobile$ = this.store.pipe(select(getIsMobileState));
  }
}
