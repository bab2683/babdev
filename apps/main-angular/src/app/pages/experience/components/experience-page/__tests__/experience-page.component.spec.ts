import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { of } from 'rxjs';

import { RequestService } from '@babdev/request';
import { RequestServiceMock } from '@babdev/request-testing';
import { TranslateService } from '@babdev/translate';
import {
  TranslatePipeMock,
  TranslateServiceMock
} from '@babdev/translate-testing';

import { AppState, GLOBAL_FEATURE } from '@store';
import { ExperiencePageComponent } from '../experience-page.component';

describe('ExperiencePageComponent', () => {
  let component: ExperiencePageComponent;
  let fixture: ComponentFixture<ExperiencePageComponent>;
  let reqService: any;
  let store: MockStore<AppState>;

  beforeEach(() => {
    reqService = new RequestServiceMock();

    TestBed.configureTestingModule({
      declarations: [ExperiencePageComponent, TranslatePipeMock],
      providers: [
        {
          provide: TranslateService,
          useClass: TranslateServiceMock
        },
        {
          provide: RequestService,
          useValue: reqService
        },
        provideMockStore({
          initialState: { [GLOBAL_FEATURE]: { isMobile: false } }
        })
      ],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ExperiencePageComponent);
    component = fixture.componentInstance;
    reqService = TestBed.inject(RequestService);
    store = TestBed.inject<Store<AppState>>(Store) as any;
    store.overrideSelector('getIsMobileState', true);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should request technologies file', () => {
    reqService.get.mockReturnValueOnce(of({ data: [] })).mockReturnValueOnce(
      of({
        data: {
          sections: {
            experiences: {
              items: [
                {
                  dates: {
                    start: '2019/08/06',
                    end: null
                  }
                },
                {
                  dates: {
                    start: '2017/08/06',
                    end: null
                  }
                },
                {
                  dates: {
                    start: '2017/08/06',
                    end: '2018/10/02'
                  }
                }
              ]
            }
          }
        }
      })
    );

    component.ngOnInit();
    expect(reqService.get).toHaveBeenCalledTimes(2);

    component.CV$.subscribe((result) => {
      expect(result.sections.experiences.items[0]).toHaveProperty('duration');
    });
  });
});
