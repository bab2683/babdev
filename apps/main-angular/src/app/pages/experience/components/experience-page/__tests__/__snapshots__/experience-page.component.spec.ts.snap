// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperiencePageComponent should match snapshot 1`] = `
<babdev-experience-page
  bgPath={[Function String]}
  dictionary={[Function Object]}
  req={[Function RequestServiceMock]}
  store={[Function MockStore]}
  translateService={[Function TranslateServiceMock]}
>
  
</babdev-experience-page>
`;
